File: C++/bfsdfs.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <list>
#include <algorithm>

const std::vector<std::string> cities = {"S", "A", "B", "C", "D", "E", "G"};
const std::vector<std::vector<int>> graph = {
    {0, 3, 5, 0, 0, 0, 0},
    {3, 0, 4, 0, 5, 0, 0},
    {5, 4, 0, 4, 0, 0, 0},
    {0, 0, 4, 0, 0, 6, 0},
    {0, 5, 0, 0, 0, 0, 3},
    {0, 0, 0, 6, 0, 0, 0},
    {0, 0, 0, 0, 3, 0, 0}
};

struct Node {
    int city;
    int cost;
    std::vector<int> path;
};

std::string formatPath(const std::vector<int>& path) {
    std::string result = "";
    for (size_t i = 0; i < path.size(); ++i) {
        result += cities[path[i]];
        if (i != path.size() - 1) {
            result += " -> ";
        }
    }
    return result;
}

int resultCost = -1;
std::string resultPath = "";

void dfs(int city, int cost, std::vector<int> path, int goal) {
    if (city == goal) {
        if (resultCost == -1 || cost < resultCost) {
            resultCost = cost;
            resultPath = formatPath(path);
        }
        return;
    }

    for (size_t next = 0; next < cities.size(); ++next) {
        if (graph[city][next] > 0 && std::find(path.begin(), path.end(), next) == path.end()) {
            std::vector<int> newPath = path;
            newPath.push_back(next);
            dfs(next, cost + graph[city][next], newPath, goal);
        }
    }
}

int main() {
    int start = 0;
    int goal = 6;

    std::queue<Node> queue;
    queue.push({start, 0, {start}});

    while (!queue.empty()) {
        Node current = queue.front();
        queue.pop();

        if (current.city == goal) {
            if (resultCost == -1 || current.cost < resultCost) {
                resultCost = current.cost;
                resultPath = formatPath(current.path);
            }
        }

        for (size_t next = 0; next < cities.size(); ++next) {
            if (graph[current.city][next] > 0 && std::find(current.path.begin(), current.path.end(), next) == current.path.end()) {
                std::vector<int> newPath = current.path;
                newPath.push_back(next);
                
                queue.push({next, current.cost + graph[current.city][next], newPath});

                dfs(next, current.cost + graph[current.city][next], newPath, goal);
            }
        }
    }

    if (resultCost != -1) {
        std::cout << "Best Path: " << resultPath << std::endl;
        std::cout << "Best Cost: " << resultCost << std::endl;
    } else {
        std::cout << "No path found." << std::endl;
    }

    return 0;
}

File: Java/bfsdfs.java
import java.util.*;

public class DFSBFSCombo {
    static String[] cities = {"S", "A", "B", "C", "D", "E", "G"};
    static int[][] graph = {
        {0, 3, 5, 0, 0, 0, 0},
        {3, 0, 4, 0, 5, 0, 0},
        {5, 4, 0, 4, 0, 0, 0},
        {0, 0, 4, 0, 0, 6, 0},
        {0, 5, 0, 0, 0, 0, 3},
        {0, 0, 0, 6, 0, 0, 0},
        {0, 0, 0, 0, 3, 0, 0}
    };

    static class Node {
        int city;
        int cost;
        List<Integer> path;
        Node(int city, int cost, List<Integer> path) {
            this.city = city;
            this.cost = cost;
            this.path = new ArrayList<>(path);
        }
    }

    public static void main(String[] args) {
        int start = 0; 
        int goal = 6;  
        String resultPath = "";
        int resultCost = Integer.MAX_VALUE;

       
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(start, 0, Arrays.asList(start)));

        while (!queue.isEmpty()) {
            Node current = queue.poll();

            if (current.city == goal && current.cost < resultCost) {
                resultCost = current.cost;
                resultPath = formatPath(current.path);
            }

            for (int next = 0; next < cities.length; next++) {
                if (graph[current.city][next] > 0 && !current.path.contains(next)) {
                    List<Integer> newPath = new ArrayList<>(current.path);
                    newPath.add(next);

                    queue.add(new Node(next, current.cost + graph[current.city][next], newPath));

                    dfs(next, current.cost + graph[current.city][next], newPath, goal);
                }
            }
        }

        System.out.println("Best Path: " + resultPath);
        System.out.println("Best Cost: " + resultCost);
    }

    static void dfs(int city, int cost, List<Integer> path, int goal) {
        if (city == goal) return;
        for (int next = 0; next < cities.length; next++) {
            if (graph[city][next] > 0 && !path.contains(next)) {
                List<Integer> newPath = new ArrayList<>(path);
                newPath.add(next);
                dfs(next, cost + graph[city][next], newPath, goal);
            }
        }
    }

    static String formatPath(List<Integer> path) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < path.size(); i++) {
            sb.append(cities[path.get(i)]);
            if (i != path.size() - 1) sb.append(" -> ");
        }
        return sb.toString();
    }
}
